File: server/internal/wealth/series_breakdown.go
package wealth

import (
	"context"
	"sort"
	"time"

	"tallyo/internal/graph/model"
	"tallyo/internal/money"

	"github.com/samber/lo"
)

var classifierSeriesOrder = []model.AssetClassifier{
	model.AssetClassifierCash,
	model.AssetClassifierPublic,
	model.AssetClassifierCompanyEquity,
	model.AssetClassifierCryptocurrency,
	model.AssetClassifierStablecoin,
	model.AssetClassifierRealEstate,
}

var liabilitySeriesOrder = []model.LiabilityCategory{
	model.LiabilityCategoryCard,
	model.LiabilityCategoryMortgage,
	model.LiabilityCategoryLoan,
	model.LiabilityCategoryOther,
}

var classifierHistoryLabels = map[model.AssetClassifier]string{
	model.AssetClassifierCash:           "Cash & Equivalents",
	model.AssetClassifierPublic:         "Public Assets",
	model.AssetClassifierCompanyEquity:  "Company Equity",
	model.AssetClassifierCryptocurrency: "Crypto",
	model.AssetClassifierStablecoin:     "Stablecoin",
	model.AssetClassifierRealEstate:     "Real Estate",
}

var liabilityHistoryLabels = map[model.LiabilityCategory]string{
	model.LiabilityCategoryCard:     "Credit Card",
	model.LiabilityCategoryMortgage: "Mortgage",
	model.LiabilityCategoryLoan:     "Loan",
}

func (s *Service) classifierSeries(ctx context.Context, input model.HistoricalNetWorthInput, dates []time.Time, tz string) ([]*model.ClassifierHistoryPoint, error) {
	if len(dates) == 0 {
		return nil, nil
	}
	startTime, endTime := seriesWindow(dates)
	raw, err := s.Store.ClassifierSnapshotValues(ctx, startTime, endTime, accountFilterFromNetWorthInput(input.Filters))
	if err != nil {
		return nil, err
	}
	byAccount := classifierSnapshotsByAccount(raw, tz)
	toPoint := func(dateKey string, classifier model.AssetClassifier, value money.Cents) *model.ClassifierHistoryPoint {
		return &model.ClassifierHistoryPoint{
			Date:       model.Date(dateKey),
			Classifier: classifier,
			Label:      labelOr(classifierHistoryLabels, classifier, string(classifier)),
			ValueUsd:   value,
		}
	}
	return historySeries(dates, byAccount, forwardFillClassifierSnapshots, classifierSeriesOrder, toPoint), nil
}

func historySeries[K comparable, S any, R any](
	dates []time.Time,
	byAccount map[int64][]S,
	forwardFill func([]S, string) map[K]money.Cents,
	order []K,
	toPoint func(string, K, money.Cents) R,
) []R {
	valuesByDate := map[string]map[K]money.Cents{}
	nonZero := map[K]bool{}
	for _, date := range dates {
		dateKey := date.Format("2006-01-02")
		valuesByDate[dateKey] = map[K]money.Cents{}
		for _, snapshots := range byAccount {
			for key, value := range forwardFill(snapshots, dateKey) {
				valuesByDate[dateKey][key] += value
			}
		}
		for key, value := range valuesByDate[dateKey] {
			if value != 0 {
				nonZero[key] = true
			}
		}
	}
	points := []R{}
	for _, date := range dates {
		dateKey := date.Format("2006-01-02")
		for _, key := range order {
			if !nonZero[key] {
				continue
			}
			points = append(points, toPoint(dateKey, key, valuesByDate[dateKey][key]))
		}
	}
	return points
}

func seriesWindow(dates []time.Time) (string, string) {
	startTime := startOfDay(dates[0]).UTC().Format(time.RFC3339)
	endTime := startOfDay(dates[len(dates)-1]).AddDate(0, 0, 1).UTC().Format(time.RFC3339)
	return startTime, endTime
}

func (s *Service) liabilitySeries(dates []time.Time, tz string, rows []SnapshotValue) []*model.LiabilityHistoryPoint {
	if len(dates) == 0 {
		return nil
	}
	raw := lo.Filter(rows, func(row SnapshotValue, _ int) bool { return isLiabilitySnapshot(row) })
	byAccount := accountSnapshotsByAccount(raw, tz)
	toPoint := func(dateKey string, category model.LiabilityCategory, value money.Cents) *model.LiabilityHistoryPoint {
		return &model.LiabilityHistoryPoint{
			Date:     model.Date(dateKey),
			Category: category,
			Label:    labelOr(liabilityHistoryLabels, category, "Other Liabilities"),
			ValueUsd: value,
		}
	}
	return historySeries(dates, byAccount, forwardFillLiabilitySnapshotValues, liabilitySeriesOrder, toPoint)
}

type classifierSnapshot struct {
	LocalDate  string
	SyncedAt   string
	Classifier model.AssetClassifier
	Value      money.Cents
	Closed     bool
}

func classifierSnapshotsByAccount(rows []ClassifierSnapshotValue, tz string) map[int64][]classifierSnapshot {
	toSnapshot := func(row ClassifierSnapshotValue) classifierSnapshot {
		return classifierSnapshot{
			LocalDate:  syncedAtToLocalDate(row.SyncedAt, tz),
			SyncedAt:   row.SyncedAt,
			Classifier: row.Classifier,
			Value:      row.ValueUSD,
			Closed:     row.AccountClosed,
		}
	}
	less := func(left, right classifierSnapshot) bool {
		if left.LocalDate != right.LocalDate {
			return left.LocalDate < right.LocalDate
		}
		if left.SyncedAt != right.SyncedAt {
			return left.SyncedAt < right.SyncedAt
		}
		return left.Classifier < right.Classifier
	}
	return snapshotsByAccount(rows, func(row ClassifierSnapshotValue) int64 { return row.AccountID }, toSnapshot, less)
}

func snapshotsByAccount[Row any, Snapshot any](
	rows []Row,
	accountID func(Row) int64,
	toSnapshot func(Row) Snapshot,
	less func(Snapshot, Snapshot) bool,
) map[int64][]Snapshot {
	toAccountSnapshot := func(row Row) (int64, Snapshot) { return accountID(row), toSnapshot(row) }
	byAccount := lo.GroupByMap(rows, toAccountSnapshot)
	for accountID := range byAccount {
		sort.Slice(byAccount[accountID], func(i, j int) bool {
			return less(byAccount[accountID][i], byAccount[accountID][j])
		})
	}
	return byAccount
}

func forwardFillClassifierSnapshots(snaps []classifierSnapshot, dateKey string) map[model.AssetClassifier]money.Cents {
	current := map[model.AssetClassifier]money.Cents{}
	for i := 0; i < len(snaps); {
		localDate, syncedAt := snaps[i].LocalDate, snaps[i].SyncedAt
		if localDate > dateKey {
			break
		}
		next := map[model.AssetClassifier]money.Cents{}
		for i < len(snaps) && snaps[i].LocalDate == localDate && snaps[i].SyncedAt == syncedAt {
			next[snaps[i].Classifier] += snaps[i].Value
			i++
		}
		current = next
	}
	if len(snaps) > 0 && snaps[0].Closed && dateKey > snaps[len(snaps)-1].LocalDate {
		// Classifier history intentionally keeps its original first-row closed check.
		return map[model.AssetClassifier]money.Cents{}
	}
	return current
}

func forwardFillLiabilitySnapshotValues(snaps []accountSnapshot, dateKey string) map[model.LiabilityCategory]money.Cents {
	category, value, ok := forwardFillLiabilitySnapshots(snaps, dateKey)
	return lo.Ternary(ok, map[model.LiabilityCategory]money.Cents{category: -value}, map[model.LiabilityCategory]money.Cents{})
}

func forwardFillLiabilitySnapshots(snaps []accountSnapshot, dateKey string) (model.LiabilityCategory, money.Cents, bool) {
	current, ok := forwardFilledSnapshot(snaps, dateKey)
	if !ok {
		return model.LiabilityCategoryOther, 0, false
	}
	return liabilitySnapshotCategory(*current), current.Value, true
}

func liabilitySnapshotCategory(snap accountSnapshot) model.LiabilityCategory {
	if snap.Manual {
		return model.LiabilityCategoryOther
	}
	return liabilityCategoryFor(snap.AccountType, snap.AccountSubtype)
}

func labelOr[K comparable](labels map[K]string, key K, fallback string) string {
	label, ok := labels[key]
	return lo.Ternary(ok, label, fallback)
}

File: server-rs/src/wealth/series_breakdown.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use chrono_tz::Tz;

use crate::{
    ids::Date,
    money::Cents,
    schema::{AssetClassifier, LiabilityCategory},
};

use super::{
    ClassifierHistoryPoint, ClassifierSnapshotValue, LiabilityHistoryPoint, SnapshotValue,
    liabilities::{self, LIABILITY_SERIES_ORDER},
    series::{AccountSnapshotValue, account_snapshots_by_account, forward_filled_snapshot},
    timezone::local_date,
};

const CLASSIFIER_SERIES_ORDER: [AssetClassifier; 6] = [
    AssetClassifier::Cash,
    AssetClassifier::Public,
    AssetClassifier::CompanyEquity,
    AssetClassifier::Cryptocurrency,
    AssetClassifier::Stablecoin,
    AssetClassifier::RealEstate,
];

pub(super) fn classifier_history_series(
    values: Vec<ClassifierSnapshotValue>,
    dates: &[DateTime<Tz>],
    timezone: &str,
) -> Vec<ClassifierHistoryPoint> {
    let by_account = classifier_snapshots_by_account(values, timezone);
    history_values(dates, &by_account, classifier_values, CLASSIFIER_SERIES_ORDER)
        .into_iter()
        .flat_map(|(date, values)| {
            CLASSIFIER_SERIES_ORDER.into_iter().filter_map(move |classifier| {
                values.get(&classifier).map(|value| ClassifierHistoryPoint {
                    date: Date::new(date.clone()).expect("sample date is canonical"),
                    classifier,
                    label: classifier_history_label(classifier).to_owned(),
                    value_usd: *value,
                })
            })
        })
        .collect()
}

pub(super) fn liability_history_series(
    values: Vec<SnapshotValue>,
    dates: &[DateTime<Tz>],
    timezone: &str,
) -> Vec<LiabilityHistoryPoint> {
    let values: Vec<_> = values
        .into_iter()
        .filter(|value| liabilities::is_liability_type(value.account_type))
        .collect();
    let by_account = account_snapshots_by_account(&values, timezone);
    history_values(dates, &by_account, liability_values, LIABILITY_SERIES_ORDER)
        .into_iter()
        .flat_map(|(date, values)| {
            LIABILITY_SERIES_ORDER.into_iter().filter_map(move |category| {
                values.get(&category).map(|value| LiabilityHistoryPoint {
                    date: Date::new(date.clone()).expect("sample date is canonical"),
                    category,
                    label: liability_history_label(category).to_owned(),
                    value_usd: *value,
                })
            })
        })
        .collect()
}

#[derive(Clone)]
struct ClassifierSnapshot {
    local_date: String,
    synced_at: DateTime<Utc>,
    classifier: AssetClassifier,
    value: Cents,
    closed: bool,
}

fn classifier_snapshots_by_account(
    values: Vec<ClassifierSnapshotValue>,
    timezone: &str,
) -> HashMap<i64, Vec<ClassifierSnapshot>> {
    let mut grouped = values.into_iter().fold(HashMap::new(), |mut grouped, value| {
        grouped
            .entry(value.account_id)
            .or_insert_with(Vec::new)
            .push(ClassifierSnapshot {
                local_date: local_date(value.synced_at, timezone),
                synced_at: value.synced_at,
                classifier: value.classifier,
                value: value.value_usd,
                closed: value.account_closed,
            });
        grouped
    });
    grouped.values_mut().for_each(|snapshots| {
        snapshots.sort_by_key(|snapshot| {
            (
                snapshot.local_date.clone(),
                snapshot.synced_at,
                snapshot.classifier.to_string(),
            )
        });
    });
    grouped
}

fn classifier_values(snapshots: &[ClassifierSnapshot], date: &str) -> HashMap<AssetClassifier, Cents> {
    let index = snapshots.partition_point(|snapshot| snapshot.local_date.as_str() <= date);
    let Some(index) = index.checked_sub(1) else {
        return HashMap::new();
    };
    let Some(snapshot) = snapshots.get(index) else {
        return HashMap::new();
    };
    if snapshots.first().is_some_and(|first| first.closed) && date > snapshots.last().unwrap().local_date.as_str() {
        return HashMap::new();
    }
    let first = snapshots[..=index]
        .iter()
        .rposition(|candidate| candidate.local_date != snapshot.local_date || candidate.synced_at != snapshot.synced_at)
        .map_or(0, |previous| previous + 1);
    snapshots[first..=index]
        .iter()
        .fold(HashMap::new(), |mut values, snapshot| {
            *values.entry(snapshot.classifier).or_default() += snapshot.value;
            values
        })
}

fn liability_values(snapshots: &[AccountSnapshotValue], date: &str) -> HashMap<LiabilityCategory, Cents> {
    let Some(snapshot) = forward_filled_snapshot(snapshots, date) else {
        return HashMap::new();
    };
    let category = if snapshot.manual {
        LiabilityCategory::Other
    } else {
        liabilities::liability_category(snapshot.account_type, snapshot.account_subtype.as_deref())
    };
    HashMap::from([(category, -snapshot.value)])
}

fn history_values<K, S>(
    dates: &[DateTime<Tz>],
    by_account: &HashMap<i64, Vec<S>>,
    values: impl Fn(&[S], &str) -> HashMap<K, Cents>,
    order: impl IntoIterator<Item = K>,
) -> Vec<(String, HashMap<K, Cents>)>
where
    K: Copy + Eq + std::hash::Hash,
{
    let order: Vec<_> = order.into_iter().collect();
    let mut non_zero = HashMap::new();
    let values_by_date: Vec<_> = dates
        .iter()
        .map(|date| {
            let date = date.format("%F").to_string();
            let values = by_account.values().fold(HashMap::new(), |mut totals, snapshots| {
                values(snapshots, &date)
                    .into_iter()
                    .for_each(|(key, value)| *totals.entry(key).or_default() += value);
                totals
            });
            values
                .iter()
                .filter(|(_, value)| **value != Cents::default())
                .for_each(|(key, _)| {
                    non_zero.insert(*key, true);
                });
            (date, values)
        })
        .collect();
    values_by_date
        .into_iter()
        .map(|(date, values)| {
            let values = order
                .iter()
                .filter(|key| non_zero.contains_key(key))
                .map(|key| (*key, *values.get(key).unwrap_or(&Cents::default())))
                .collect();
            (date, values)
        })
        .collect()
}

fn classifier_history_label(classifier: AssetClassifier) -> &'static str {
    match classifier {
        AssetClassifier::Cash => "Cash & Equivalents",
        AssetClassifier::Public => "Public Assets",
        AssetClassifier::CompanyEquity => "Company Equity",
        AssetClassifier::Cryptocurrency => "Crypto",
        AssetClassifier::Stablecoin => "Stablecoin",
        AssetClassifier::RealEstate => "Real Estate",
    }
}

fn liability_history_label(category: LiabilityCategory) -> &'static str {
    match category {
        LiabilityCategory::Card => "Credit Card",
        LiabilityCategory::Mortgage => "Mortgage",
        LiabilityCategory::Loan => "Loan",
        LiabilityCategory::Other => "Other Liabilities",
    }
}

#[cfg(test)]
mod tests {
    use chrono::TimeZone;

    use super::*;

    #[test]
    fn classifier_history_keeps_zero_points_after_a_classifier_has_value() {
        let dates = [
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 1, 12, 0, 0).unwrap(),
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 2, 12, 0, 0).unwrap(),
        ];
        let points = classifier_history_series(
            vec![
                ClassifierSnapshotValue {
                    account_id: 1,
                    account_closed: false,
                    synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                    classifier: AssetClassifier::Cash,
                    value_usd: Cents(100),
                },
                ClassifierSnapshotValue {
                    account_id: 1,
                    account_closed: false,
                    synced_at: "2026-06-02T12:00:00Z".parse().unwrap(),
                    classifier: AssetClassifier::Cash,
                    value_usd: Cents::default(),
                },
            ],
            &dates,
            "UTC",
        );

        assert_eq!(
            points.iter().map(|point| point.value_usd).collect::<Vec<_>>(),
            [Cents(100), Cents::default()]
        );
    }

    #[test]
    fn classifier_and_liability_history_group_current_snapshot_values() {
        let dates = [
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 1, 12, 0, 0).unwrap(),
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 2, 12, 0, 0).unwrap(),
        ];
        let classifiers = classifier_history_series(
            vec![
                ClassifierSnapshotValue {
                    account_id: 1,
                    account_closed: false,
                    synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                    classifier: AssetClassifier::Cash,
                    value_usd: Cents(100),
                },
                ClassifierSnapshotValue {
                    account_id: 1,
                    account_closed: false,
                    synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                    classifier: AssetClassifier::Public,
                    value_usd: Cents(200),
                },
                ClassifierSnapshotValue {
                    account_id: 1,
                    account_closed: false,
                    synced_at: "2026-06-02T12:00:00Z".parse().unwrap(),
                    classifier: AssetClassifier::Cash,
                    value_usd: Cents(50),
                },
            ],
            &dates,
            "UTC",
        );
        let liabilities = liability_history_series(
            vec![
                SnapshotValue {
                    account_id: 1,
                    account_type: crate::accounts::AccountType::Credit,
                    account_manual: false,
                    account_subtype: None,
                    account_closed: false,
                    synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                    balance_usd: Cents(-300),
                },
                SnapshotValue {
                    account_id: 2,
                    account_type: crate::accounts::AccountType::Loan,
                    account_manual: true,
                    account_subtype: Some("mortgage".to_owned()),
                    account_closed: false,
                    synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                    balance_usd: Cents(-400),
                },
            ],
            &dates,
            "UTC",
        );

        assert_eq!(
            classifiers.iter().map(|point| point.value_usd).collect::<Vec<_>>(),
            [Cents(100), Cents(200), Cents(50), Cents::default()]
        );
        assert_eq!(
            liabilities.iter().map(|point| point.value_usd).collect::<Vec<_>>(),
            [Cents(300), Cents(400), Cents(300), Cents(400)]
        );
        assert_eq!(classifier_history_label(AssetClassifier::Cryptocurrency), "Crypto");
        assert_eq!(liability_history_label(LiabilityCategory::Mortgage), "Mortgage");
    }

    #[test]
    fn labels_every_history_series_and_closes_classifier_history() {
        let dates = [
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 1, 12, 0, 0).unwrap(),
            chrono_tz::UTC.with_ymd_and_hms(2026, 6, 2, 12, 0, 0).unwrap(),
        ];
        let points = classifier_history_series(
            vec![ClassifierSnapshotValue {
                account_id: 1,
                account_closed: true,
                synced_at: "2026-06-01T12:00:00Z".parse().unwrap(),
                classifier: AssetClassifier::Cash,
                value_usd: Cents(100),
            }],
            &dates,
            "UTC",
        );

        assert_eq!(
            [
                AssetClassifier::Cash,
                AssetClassifier::Public,
                AssetClassifier::CompanyEquity,
                AssetClassifier::Cryptocurrency,
                AssetClassifier::Stablecoin,
                AssetClassifier::RealEstate,
            ]
            .map(classifier_history_label),
            [
                "Cash & Equivalents",
                "Public Assets",
                "Company Equity",
                "Crypto",
                "Stablecoin",
                "Real Estate",
            ]
        );
        assert_eq!(
            [
                LiabilityCategory::Card,
                LiabilityCategory::Mortgage,
                LiabilityCategory::Loan,
                LiabilityCategory::Other,
            ]
            .map(liability_history_label),
            ["Credit Card", "Mortgage", "Loan", "Other Liabilities"]
        );
        assert_eq!(
            points.iter().map(|point| point.value_usd).collect::<Vec<_>>(),
            [Cents(100), Cents::default()]
        );
    }
}
